### DISCUSSION TITLE
August 2022 Roadmap

### DISCUSSION THREAD

**Discussion by tjcran** [coder/coder#3328] [RFCs]
Aug 8 - Sept 2
### 🛂 Enterprise Licensing

This represents the work needed to allow for license generation and "hiding" enterprise features behind a locked license gate. This is for both frontend and backend-related work. 

- [ ] #480 @spikecurtis @presleyp 

### 🔐 OIDC 
This will likely be a generic OIDC implementation, similar to the v1 Coder product. However, some considerations need to be made for how we could make this extensible to support advanced auth use cases in the future, like SAML support. 

- [x] #3314 @kylecarbs 

### 🕐  Workspace Schedule Lifecycle

Details TBD, being fleshed out by @bpmct 

- [ ] #2995 @bpmct 

### 🕵️‍♀️ Audit Logging 
Audit Logging will be the first released "Enterprise" paid feature. This feature will allow for admins to track events and actions in Coder to maintain security, diagnose and troubleshoot, and better understand usage. This will include both UI, CLI, and exporting via `slog` capabilities. 

- [ ] #3251 @coadler @Kira-Pilot 

### 🧊 Helm Chart

For users who want to run their Coder dashboard and workspaces in Kubernetes, a Helm chart install should allow users to install and upgrade their Coder deployment. The first pass will allow for a simple deployment with just a Coder dashboard, but subsequent helm charts should include HA options and the ability to run containerized workspaces in the same K8s cluster.

- [x] #2267 @deansheather 
- [ ] #3265 @deansheather 

### 🔌 External Provisioner Daemons

Connect (and authenticate) external environments for provisioning infrastructure outside of the server running Coder. For example, `coder server` can run seperately from the machine(s) that are responsible for provisioning workspaces. See our [architecture](https://coder.com/docs/coder-oss/latest/architecture#provisionerd) for more info.

- [ ] #44 @dwahler 

#### 💯 Must have

- [ ] #802 / #2660
- [ ] #2779
- [x] #2135 
- [x] #2989 
- [ ] #2662 
- [x] https://github.com/coder/coder/issues/3051 / https://github.com/coder/coder/issues/2960
- [x] https://github.com/coder/coder/issues/2874
- [x] #2778
- [ ] https://github.com/coder/coder/issues/3378
- [x] https://github.com/coder/coder/issues/3223 / https://github.com/coder/coder/issues/3225
- [ ] https://github.com/coder/coder/issues/3309
- [ ] https://github.com/coder/coder/issues/3304

#### 🎨 Nice to have

- [x] #2659
- [ ] #2221
- [ ] #3054
- [x] #2573
- [ ] https://github.com/coder/coder/issues/3005
- [ ]  https://github.com/coder/coder/issues/2764
- [ ]  https://github.com/coder/coder/issues/2752
- [ ]  https://github.com/coder/coder/issues/2970
- [ ]  https://github.com/coder/coder/issues/3000
- [x]  https://github.com/coder/coder/issues/2714
- [x]  https://github.com/coder/coder/issues/1369
- [x]  https://github.com/coder/coder/issues/2836
- [x]  https://github.com/coder/coder/issues/2146
- [x]  https://github.com/coder/coder/issues/3013
- [ ]  https://github.com/coder/coder/issues/2966
- [x]  https://github.com/coder/coder/issues/2842
- [ ]  https://github.com/coder/coder/issues/2840
- [x]  https://github.com/coder/coder/issues/3031
- [ ]  https://github.com/coder/coder/issues/2815

**Comment by ammario** [comment]
The "no more god mode" (https://github.com/coder/coder/issues/2135) should be counted as a smaller must-have item. It should take a few hours at most unless I'm missing something about our implementation.

**Comment by ammario** [comment]
https://github.com/coder/coder/issues/2995 is worth consideration. It's probably amongst our top 3 feature requests in terms of feedback.

**Comment by ammario** [comment]
Note to carry over Resource Metadata and template.yaml from the July roadmap (#3042) that almost certainly will not be finished.

**Reply by bpmct** [reply]
@BrunoQuaresma @dwahler - I put #2989 in the "smaller, but mighty" section since it's almost done and addressed in the July roadmap.

**Comment by ammario** [comment]
Is Frontend Edit Workspace (https://github.com/coder/coder/issues/802) really a big feature? The issue describes a page with simple forms and we already have the backend endpoints.

**Comment by tjcran** [comment]
> Is Frontend Edit Workspace really a big feature? The issue describes a page with simple forms and we already have the backend endpoints.

@ammario i think the misalignment is that we don't have a rubric for when something is big enough to be considered a "major" item on the roadmap. I think you're right, though. This one should not be on the "major" list.

**Comment by ammario** [comment]
@tjcran a heuristic I use is whether we would include the feature in a release blog post. There's still a lot of room for interpretation, but perhaps slightly less.

**Comment by spikecurtis** [comment]
@tjcran @kylecarbs I feel like I'm seeing a pattern of bugs related to orphaned resources in the cloud provider.  It's pretty concerning because it can cause wasted cloud spend.  Our architecture mostly trusts the provisioner (terraform) to clean up after itself, but there are clearly edge cases like cancellation of jobs, bugs in terraform providers, etc.  I wonder whether we need to think about some sort of reconciliation process.

**Reply by bpmct** [reply]
Putting all the replies in here to make the conversation a bit easier to follow:

> @tjcran:
> This sounds like it needs an epic to wrap all these together? Is there a need for an RFC on this to consider a more holistic strategy for dealing with it?
> Also noting https://github.com/coder/coder/issues/3048 as related

> @spikecurtis:
> I'd want to start with some strategic discussion & brainstorming, maybe @ketang @kylecarbs @spikecurtis @tjcran (optionally @ammario, optionally other engineers with deep distributed systems experience). One outcome of that discussion could be to move to RFC, but it could also be that we're going to punt on the larger problem and just fix bugs.

> @ammario:
> @spikecurtis is this RFC something you are proposing for the August roadmap?

> @spikecurtis:
> My personal feeling is that this trumps several things on the August list, like activity based auto-off.

> @ammario 
> I have encountered the problem personally and I agree it's a fundamental problem with our model that we need to resolve. We have a few issues related to this problem, and no silver bullet has been identified. There are a bunch of small things that make progress, e.g https://github.com/coder/coder/issues/3035 and https://github.com/coder/coder/issues/2256#issuecomment-1152740102.
> 
> But it's not clear what exactly we would pursue on if the "orphanage problem" was added to roadmap.
> 
> You can probably think of more/better solutions than the aforementioned issues too.
> 
> Another thing to note is activity based auto stop has been requested many times by users where orphanage resolution hasn't.

> @tjcran 
> I suggest we add the orphaning problem to the roadmap as an RFC item for August so we can work out how to solve it. But I don't think we can properly assess the cost and prioritize it to this month's roadmap without an RFC.

**Comment by tjcran** [comment]
> @tjcran @kylecarbs I feel like I'm seeing a pattern of bugs related to orphaned resources in the cloud provider. It's pretty concerning because it can cause wasted cloud spend. Our architecture mostly trusts the provisioner (terraform) to clean up after itself, but there are clearly edge cases like cancellation of jobs, bugs in terraform providers, etc. I wonder whether we need to think about some sort of reconciliation process.

This sounds like it needs an epic to wrap all these together? Is there a need for an RFC on this to consider a more holistic strategy for dealing with it?

Also noting https://github.com/coder/coder/issues/3048 as related

**Comment by spikecurtis** [comment]
I'd want to start with some strategic discussion & brainstorming, maybe @ketang @kylecarbs @spikecurtis @tjcran (optionally @ammario, optionally other engineers with deep distributed systems experience).  One outcome of that discussion could be to move to RFC, but it could also be that we're going to punt on the larger problem and just fix bugs.

**Comment by tjcran** [comment]
> I'd want to start with some strategic discussion & brainstorming, maybe @ketang @kylecarbs @spikecurtis @tjcran (optionally @ammario, optionally other engineers with deep distributed systems experience). One outcome of that discussion could be to move to RFC, but it could also be that we're going to punt on the larger problem and just fix bugs.

@spikecurtis is this RFC something you are proposing for the August roadmap?

**Comment by spikecurtis** [comment]
My _personal_ feeling is that this trumps several things on the August list, like activity based auto-off.

**Comment by ammario** [comment]
> My _personal_ feeling is that this trumps several things on the August list, like activity based auto-off.

I have encountered the problem personally and I agree it's a fundamental problem with our model that we need to resolve. We have a few issues related to this problem, and no silver bullet has been identified. There are a bunch of small things that make progress, e.g  https://github.com/coder/coder/issues/3035 and https://github.com/coder/coder/issues/2256#issuecomment-1152740102.

But it's not clear what exactly we would pursue on if the "orphanage problem" was added to roadmap. 

You can probably think of more/better solutions than the aforementioned issues too.

Another thing to note is activity based auto stop has been requested many times by users where orphanage resolution hasn't.

**Comment by tjcran** [comment]
I suggest we add the orphaning problem to the roadmap as an _RFC item_ for August so we can work out how to solve it. But I don't think we can properly assess the cost and prioritize it to this month's roadmap without an RFC.

**Comment by presleyp** [comment]
Just something to be aware of: Licensing and Audit Logging aren't totally separable. They each have independent tasks to start with, but when it's time to add audit logging to the app, it's also time to gate audit logging by license. Seen from the other side, we can't fully implement licensing without a conditional feature to implement it on.

#1763 should be replaced by the first of the two tickets replacing it (the second is less urgent so it's up to you whether it makes it on the August roadmap)

I'm not clear on the requirements for #802 / #2660. #802 was going to be a page you can navigate to whenever you want to change params. #2660 sounds like it's asking for the app to check for template changes when the user clicks Update, and if they exist, display a form to fill in those inputs.

**Reply by bpmct** [reply]
Moving conversation into a thread:

>  @tjcran 
> good callout. I'll make the update for https://github.com/coder/coder/issues/1763 item.
The workspace edit issue needs some prep work and some product direction. @bpmct were you still planning to work on that one?

> @bpmct 
> #802 and #660 is something that we need to define more in August! However, it is still something we want to do in August. It needs an owner and I'd estimate it's fairly complex and needs backend work as well.

**Reply by bpmct** [reply]
I added a problem statement and definition of done to #802, it is now a lot more than a page to edit parameters, as there is some underlying complexity that needs to be worked out first.

I think it will likely become an epic and spin out into a few different tasks, once a solution is landed on.

**Comment by tjcran** [comment]
> Just something to be aware of: Licensing and Audit Logging aren't totally separable. They each have independent tasks to start with, but when it's time to add audit logging to the app, it's also time to gate audit logging by license. Seen from the other side, we can't fully implement licensing without a conditional feature to implement it on.
> 
> #1763 should be replaced by the first of the two tickets replacing it (the second is less urgent so it's up to you whether it makes it on the August roadmap)
> 
> I'm not clear on the requirements for #802 / #2660. #802 was going to be a page you can navigate to whenever you want to change params. #2660 sounds like it's asking for the app to check for template changes when the user clicks Update, and if they exist, display a form to fill in those inputs.

@presleyp good callout. I'll make the update for #1763 item. 
The workspace edit issue needs some prep work and some product direction. @bpmct were you still planning to work on that one?

**Comment by bpmct** [comment]
@presleyp #802 and #2660 is something that we need to define more in August! However, it is still something we want to do in August. It needs an owner and I'd estimate it's fairly complex and needs backend work as well.

**Comment by ammario** [comment]
For enterprise, https://github.com/coder/coder/issues/3304 is a good simple initial feature and we've already had a prospect request it.

**Comment by ammario** [comment]
[Here](https://github.com/coder/coder/issues/3309) is another simple yet urgent issue since it will help us better understand this early usage and make a better roadmap in September.

**Comment by bpmct** [comment]
Suggesting we add #2662 as well as a nice-to-have.

Edit: or even a must-have. A lot of our docs will run off scripts to install code-server or projector, and health checks help avoid broken experiences for users before the IDE loads.

**Reply by sharkymark** [reply]
Opening IDEs and getting errors seems like basic error handling and should be a high priority. 

I reported this in June.

**Reply by tjcran** [reply]
Which IDEs are still seeing issues? @spikecurtis did a lot of work on JetBrains as part of the July roadmap #3042

**Reply by spikecurtis** [reply]
I did work on JetBrains Gateway, which doesn't use `coder_apps`

**Reply by tjcran** [reply]
Good point. Either way, I think the issue Ben suggested should be a high prio this month.

**Comment by sharkymark** [comment]
Adding a dotfiles repo at the account level is parity with v1. #3114

**Reply by tjcran** [reply]
This one should be part of the September roadmap

**Comment by bpmct** [comment]
Where does Tailscale (#2779) and affected issues (#3316) land in the August roadmap? @kylecarbs I was able to reproduce #3116 on remote (non-Docker) workspaces.

**Reply by kylecarbs** [reply]
I'll be working on getting this merged in today, probably under a flag. If we see no issues after dogfooding this week, I'd say it's good to make primary.

**Reply by bpmct** [reply]
Cool - I'll add it under "must have" for August.

**Comment by bpmct** [comment]
We've seen some confusion about the current `coder templates pull` behavior from 2 users. I think it would be nice to address as a nice-to-have: #2651

**Comment by bpmct** [comment]
#2778 as a must have. The agent switching orders on the dashboard is pretty unprofessional

**Reply by bpmct** [reply]
I went ahead and just added it.
